crypto/sm3/asm/sm3-x86_64.pl

```diff
@@ -0,0 +1,344 @@
+    vpalignr        \$8, $W11_08, $W15_12, $T2
+    vsm3msg2        $T2, $T1, $T3
+___
+}
+
+# Performs 4 rounds of SM3 algorithm
+#   - consumes 4 words of message schedule W[]
+#   - updates SM3 state registers: ABEF and CDGH
+sub sm3rounds4 {
+my ($ABEF, $CDGH, $W03_00, $W07_04, $T1,$R)=@_;
+my $R2 = $R + 2;
+$code.=<<___;
+    vpunpcklqdq     $W07_04, $W03_00, $T1
+    vsm3rnds2       \$$R, $T1, $ABEF, $CDGH
+    vpunpckhqdq     $W07_04, $W03_00, $T1
+    vsm3rnds2       \$$R2, $T1, $CDGH, $ABEF
+___
+}
+
+$code.= ".data\n";
+{
+# input arguments aliases
+my ($ctx,$p,$num) = @_3args;
+
+$code.=<<___;
+.align 16
+SHUFF_MASK:
+    .byte 3, 2, 1, 0, 7, 6, 5, 4, 11, 10, 9, 8, 15, 14, 13, 12
+
+.text
+
+# void ossl_hwsm3_block_data_order(SM3_CTX *c, const void *p, size_t num)
+#
+# input: $ctx SM3 context
+#        $p  pointer to the data
+#        $num number of blocks
+#
+
+.globl	ossl_hwsm3_block_data_order
+.type	ossl_hwsm3_block_data_order,\@function,3
+.align	32
+ossl_hwsm3_block_data_order:
+.cfi_startproc
+    endbranch
+# Prolog
+    push    %rbp
+.cfi_push   %rbp
+.Lossl_hwsm3_block_data_order_seh_push_rbp:
+    # ; %rbp contains stack pointer right after GP regs pushed at stack + [8
+    # ; bytes of alignment (Windows only)].  It serves as a frame pointer in SEH
+    # ; handlers. The requirement for a frame pointer is that its offset from
+    # ; RSP shall be multiple of 16, and not exceed 240 bytes. The frame pointer
+    # ; itself seems to be reasonable to use here, because later we do 64-byte stack
+    # ; alignment which gives us non-determinate offsets and complicates writing
+    # ; SEH handlers.
+    #
+    # ; It also serves as an anchor for retrieving stack arguments on both Linux
+    # ; and Windows.
+    lea     `$XMM_STORAGE`(%rsp),%rbp
+.cfi_def_cfa_register %rbp
+.Lossl_hwsm3_block_data_order_seh_setfp:
+___
+  if ($win64) {
+
+    # ; xmm6:xmm15 need to be preserved on Windows
+    foreach my $reg_idx (6 .. 15) {
+      my $xmm_reg_offset = ($reg_idx - 6) * 16;
+      $code .= <<___;
+        vmovdqu           %xmm${reg_idx},$xmm_reg_offset(%rsp)
+.L${func_name}_seh_save_xmm${reg_idx}:
+___
+    }
+  }
+
+  $code .= <<___;
+# Prolog ends here.
+.Lossl_hwsm3_block_data_order_seh_prolog_end:
+    or $num, $num
+    je .done_hash
+
+    vmovdqu         ($ctx), %xmm6
+    vmovdqu         16($ctx), %xmm7
+
+    vpshufd         \$0x1B, %xmm6, %xmm0
+    vpshufd         \$0x1B, %xmm7, %xmm1
+    vpunpckhqdq     %xmm0, %xmm1, %xmm6
+    vpunpcklqdq     %xmm0, %xmm1, %xmm7
+    vpsrld          \$9, %xmm7, %xmm2
+    vpslld          \$23, %xmm7, %xmm3
+    vpxor           %xmm3, %xmm2, %xmm1
+    vpsrld          \$19, %xmm7, %xmm4
+    vpslld          \$13, %xmm7, %xmm5
+    vpxor           %xmm5, %xmm4, %xmm0
+    vpblendd        \$0x3, %xmm0, %xmm1, %xmm7
+
+    vmovdqa         SHUFF_MASK(%rip), %xmm12
+
+.align 32
+.block_loop:
+    vmovdqa         %xmm6, %xmm10
+    vmovdqa         %xmm7, %xmm11
+
```

Comment:
It may be worth to add a note about register content comment notation
```
xmm = A B C D
A - most significant word in an `xmm`
D - least significant word in an `xmm`
```

Reply:
Thank you, Tomasz, applied, just propose to reor this one as:
```
    # xmm = D C B A
    # D - most significant word in an `xmm`
    # A - least significant word in an `xmm`
```
to match notion in `xmm6 = D C B A, xmm7 = H G F E` https://github.com/openssl/openssl/pull/26196/commits/6cff5a5506f75dabb097b051570f7b607a45e44b